### PR TITLE
Rules

### PR DESCRIPTION
**Pull request by dantleech**
This PR introduces the linter and 2 rules:

- `no-empty-file`: Disallow empty files
- `no-duplicate-tags`: Disallow duplicate tags

image

### CODE REVIEW

File: src/GherkinLintContainer.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    private function createFinder(string $cwd): FeatureFinder
+    {
+        return new FeatureFinder($cwd);
+    }
+
+    private function createLinter(): Linter
+    {
+        return new Linter(
+            new GherkinParser(
+                includeSource: false,
```
Comment (dantleech) [inline src/GherkinLintContainer.php:38]: should probably create a factory for this (currently we depend on NOT including source)

File: src/Model/Linter.php [hunk 2]
```diff
@@ -1,19 +1,40 @@
...
-    public function lint(string $uri, string $contents): FeatureDiagnostics
+    /**
+     * @return Generator<FeatureDiagnostic>
+     */
+    public function lint(string $uri, string $contents): Generator
     {
-        $node = $this->parser->parseString($uri, $contents);
+        $envelopes = iterator_to_array($this->parser->parseString($uri, $contents));
+
+        foreach ($envelopes as $envelope) {
```
Comment (dantleech) [inline src/Model/Linter.php:26]: @ciaranmcnulty not really sure why there are multiple envelopes? is it expected to pass `includeSource`, `includeGherkinDocument`, `includePickles` (what are Pickles?) to the parser to determine what is returned?

File: src/Model/Linter.php [hunk 3]
```diff
@@ -1,19 +1,40 @@
...
+            foreach ($this->rules as $rule) {
+                $document = $envelope->gherkinDocument;
+                if (null === $document) {
```
Comment (dantleech) [inline src/Model/Linter.php:29]: the rules currently only get the GherkinDocument, is there any value in passing in other things?

File: src/Model/Linter.php [hunk 4]
```diff
@@ -1,19 +1,40 @@
...
+                    continue;
+                }
+
+                yield from $rule->analyse($document);
+            }
 
-        return new FeatureDiagnostics([]);
+            // why multiple envelopes?
+            break;
```
Comment (dantleech) [inline src/Model/Linter.php:37]: remove this and possibly ensure that we only get the "gherkin document" envelope

File: src/Model/Linter.php (hunk 2 above)
Comment (ciaranmcnulty) [reply src/Model/Linter.php:26]: Messages are used for more than just gherkin, it's also the test execution events (see https://github.com/cucumber/common/tree/main/messages#message-overview) - ignore any that aren't relevant

Just for clarity though here's what the parser can emit:

 * Source - this is the raw `.feature` file and its URL. You would probably only want this if you want to output a chunk of the file in output? It's more convenient than recreating from the AST
 * GherkinDocument - this is the AST, you probably want this
 * Pickle - this is the AST normalised into something like `list<list<Step>>` - it's for a test runner to not have to worry about things like Background. You aren't likely to want this

Comment (ciaranmcnulty) [reply src/Model/Linter.php:26]: (basically envelope is so we can naively decode JSON without knowing out-of-band what 'type' it is, then look at its fields to see what it was)

Comment (dantleech) [reply src/Model/Linter.php:26]: ok so could "filter" envelopes by checking if `gherkinDocument` is NULL or not I guess

**Comment by ciaranmcnulty** [comment]
I like laziness so I would probably:

* Make FeatureFinder::find return Generator
* let Linter::lint() accept iterable<SplFileInfo> as an argument

Then the code in LintCommand would look like:

```php
$files = $this->finder->find($path);
$diagnostics = $this->linter->lint($files);
```

The Linter would be more like:

```php
public function lint($files) 
{
    $sources = $this->sources($files);
    $envelopes = $this->parser->parse($sources);
    $documents = $this->gherkinDocuments($envelopes);
    
    yield from $this->analyse($documents);
}

private function sources($files)
{
    foreach ($files as $file) {
        yield new Source(uri: $file->pathName(), data: file_get_contents($fileInfo->path));
    }
}

private function gherkinDocuments($envelopes) 
{
    foreach ($envelopes as $envelope) {
        if (!is_null($envelope->gherkinDocument)) {
            yield $envelope->gherkinDocument;
        }
    }
}

private function analyse($documents)
{
    foreach ($documents as $document) {
        foreach ($this->rules as $rule) {
            yield from $rule->analyse($document);
        }
    }
}

```

**Comment by dantleech** [comment]
> Make FeatureFinder::find return Generator

i was going to do this, however if we want a progress bar then we need to count the files anyway... so makes no difference (assuming a progress bar is useful).

otherwise your Linter looks better ....

**Comment by ciaranmcnulty** [comment]
That's why dot formatters are so much easier :)

**Comment by dantleech** [comment]
Not cool though

**Comment by dantleech** [comment]
antother thing is that Diagnostics need to be retrieved on a per-file basis to be able to organise them in report, which is why the Linter accepts a single file

**Comment by dantleech** [comment]
ok, i think I'll refactor to use Generators throughout, I'll merge this now though,

**Comment by dantleech** [comment]
Refactored a bit here: https://github.com/dantleech/gherkin-lint-php/pull/2

**Pull request merged**